src/python/Utils/TokenManager.py

```diff
@@ -0,0 +1,130 @@
+        return token
+    if name:
+        return os.environ.get(name)
+    return os.environ.get("IAM_TOKEN")
+
+def tokenData(token, url=None, aud=None):
+    """
+    inspect and extract token data
+    :param token: token string
+    :param url: IAM provider URL, by default it is https://cms-auth.web.cern.ch/jwk
+    :param aud: audience string, by default it is https://wlcg.cern.ch/jwt/v1/any
+    """
+    if not token or not jwt:
+        return {}
+    if isinstance(token, str):
+        token = bytes(token, 'utf-8')
+    if not url: # use CMS IAM
+        url = "https://cms-auth.web.cern.ch/jwk"
+    jwksClient = jwt.PyJWKClient(url)
+    signingKey = jwksClient.get_signing_key_from_jwt(token)
+    key = signingKey.key
+    headers = jwt.get_unverified_header(token)
+    alg = headers.get('alg', 'RS256')
+    if not aud: # use WLCG
+        aud = "https://wlcg.cern.ch/jwt/v1/any"
+    data = jwt.decode(
+        token,
+        key,
+        algorithms=[alg],
+        audience=aud,
+        options={"verify_exp": True},
+    )
+    return data
+
+def validToken(token):
+    """
+    validate token expiration timestamp
+    :param token: token string
+    """
+    tokenDict = {}
+    try:
+        tokenDict = tokenData(token)
+    except:
+        traceback.print_exc()
+        raise
+    exp = tokenDict.get('exp', 0)
+    if not exp or exp < time.time():
+        return False
+    return True
+
+
+class TokenManager():
+    """
+    TokenManager class handles IAM tokens
+    """
+    def __init__(self, name=None, url=None, audUrl=None, logger=None):
+        """
+        Token manager reads IAM tokens either from file or env.
+        It caches token along with expiration timestamp.
+        By default the env variable to use is IAM_TOKEN.
+        :param name: string representing either file or env where we should read token from
+        :param url: IAM provider URL, by default it is https://cms-auth.web.cern.ch/jwk
+        :param audUrl: audience string, by default it is https://wlcg.cern.ch/jwt/v1/any
+        :param logger: logger object or none to use default one
+        """
+        self.name = name
+        self.url = url
+        self.aud = audUrl
+        self.expire = 0
+        self.token = None
+        self.logger = logger if logger else logging.getLogger()
+        try:
+            self.token = self.getToken()
+        except Exception as exc:
+            self.logger.exception(str(exc))
+
+    def getToken(self):
+        """
+        Return valid token and sets its expire timestamp
+        """
+        if not self.token or not validToken(self.token):
+            self.token = getToken(self.name)
+        tokenDict = {}
+        try:
+            tokenDict = tokenData(self.token, url=self.url, aud=self.aud)
+            self.logger.debug(tokenDict)
+        except Exception as exc:
+            self.logger.exception(str(exc))
+            raise
+        self.expire = tokenDict.get('exp', 0)
```

Comment:
Along the comment made above - for creating a new function - now that I look into this, it makes sense to me to have a method called `getExpirationTime` or `getLifetime` to read this attribute.

Reply:
no, it should be `isValidToken` since we read token and it can be mailformed, and only if we read it correctly we verify its expiration timestamp. So, let's not change it.

Reply:
Yes, what I wanted to say is that we should have something like:
```
def getLifetime(self):
    return self.expire - int(time.time()) 
```

The use case here is that we want to renew this token BEFORE it gets expired. So we need to have a mechanism to constantly poll how long it's still valid for.

Reply:
Resolving this as well, since you implemented this method below. Thanks